### PR TITLE
KAFKA-10262: Ensure that creating task directory is thread safe

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @ableegoldman @vvcephei

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java [hunk 1]
```diff
@@ -507,4 +510,51 @@ public void shouldLockGlobalStateDirectoryWhenDirectoryCreationDisabled() throws
         initializeStateDirectory(false);
         assertTrue(directory.lockGlobalState());
     }
+
+    @Test
+    public void shouldNotFailWhenCreatingTaskDirectoryInParallel() throws Exception {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:515]: Without the fix, this test fails like 100% of the time for me. (Kinda surprising -- seems that the disk IO operations give us nice thread context switches so we hit the issue reliably.)

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:515]: Huh, neat!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 2]
```diff
@@ -47,10 +47,10 @@
 public class StateDirectory {
 
     private static final Pattern PATH_NAME = Pattern.compile("\\d+_\\d+");
-
-    static final String LOCK_FILE_NAME = ".lock";
     private static final Logger log = LoggerFactory.getLogger(StateDirectory.class);
+    static final String LOCK_FILE_NAME = ".lock";
 
+    private final Object taskCreationLock = new Object();
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: super nit: taskCreationLock --> taskDirCreationLock

**Review by vvcephei — APPROVED** [review]
Thanks for the fix, @mjsax !

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateDirectoryTest.java:515]: In retrospect, this makes sense. The fs operations probably take an order of magnitude (at least) more time than the difference in time between starting the two threads. Thanks for adding it!

**Review by ableegoldman — APPROVED** [review]
Thanks @mjsax! One pedantic nit but otherwise this LGTM.

Not sure where we are with the 2.6 release or any RCs being cut, but we probably need to ask/inform Randall about merging this fix to 2.6

**Comment by ableegoldman** [comment]
Oh nevermind, you literally just did. Ignore my last 🙂

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: That's fair.

**Comment by mjsax** [comment]
Two failures (J11 and J14) of
```
org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testOneWayReplicationWithAutorOffsetSync1
```

J8 passed.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.6`.

**Pull request merged**